pkg/controller/storagecluster/storagecluster.go

```diff
@@ -1176,12 +1183,44 @@ func (c *Controller) getCurrentMaxStorageNodesPerZone(
 	return 0, fmt.Errorf("storage disabled")
 }
 
+func getDefaultStorageNodesDisaggregatedMode(nodeList *v1.NodeList, cluster *corev1.StorageCluster) (uint64, error) {
+	// Check if ENABLE_ASG_STORAGE_PARTITIONING is set to false. If not, we can look at the 'portworx.io/node-type' label
+	for _, envVar := range cluster.Spec.Env {
+		if envVar.Name == StoragePartitioningEnvKey {
+			if envVar.Value == "false" {
+				return 0, fmt.Errorf("%v is set to false", StoragePartitioningEnvKey)
+			}
+			break
+		}
+	}
+
+	// This will return a zone map of storage nodes in each zones
+	nodeTypeZoneMap := getZoneMap(nodeList, NodeTypeKey, StorageNodeValue)
+	// We'll find the zone with least number of labels
+	minValue := uint64(math.MaxUint64)
+	for _, value := range nodeTypeZoneMap {
+		if value < minValue {
+			minValue = value
+		}
+	}
+	return minValue, nil
+}
+
 // getDefaultMaxStorageNodesPerZone aims to return a good value for MaxStorageNodesPerZone with the
 // intention of having at least 3 nodes in the cluster.
-func getDefaultMaxStorageNodesPerZone(nodeList *v1.NodeList) (uint32, error) {
-	zoneMap := getZoneMap(nodeList)
+func getDefaultMaxStorageNodesPerZone(nodeList *v1.NodeList, cluster *corev1.StorageCluster) (uint32, error) {
+	var storageNodes uint64
+	var err error
+
+	// Check if storage nodes are explicitly set using labels
+	storageNodes, err = getDefaultStorageNodesDisaggregatedMode(nodeList, cluster)
+	if err == nil {
```

Comment:
this does not handle unexpected error, we can use a bool to indicate whether it's disaggregated mode or not.